libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/RustClientSessionDelegate.kt

```diff
@@ -0,0 +1,124 @@
+                "Saving new session data with token: access token '$anonymizedAccessToken' and refresh token '$anonymizedRefreshToken'. " +
+                    "Was token valid: ${existingData.isTokenValid}"
+            )
+            val newData = session.toSessionData(
+                isTokenValid = true,
+                loginType = existingData.loginType,
+                passphrase = existingData.passphrase,
+                sessionPaths = existingData.getSessionPaths(),
+            )
+            sessionStore.updateData(newData)
+            clientLog.d("Saved new session data with access token: '$anonymizedAccessToken'.")
+        }.invokeOnCompletion {
+            if (it != null) {
+                clientLog.e(it, "Failed to save new session data.")
+            }
+        }
+    }
+
+    override fun didReceiveAuthError(isSoftLogout: Boolean) {
+        clientLog.w("didReceiveAuthError(isSoftLogout=$isSoftLogout)")
+        if (isLoggingOut.getAndSet(true).not()) {
+            clientLog.v("didReceiveAuthError -> do the cleanup")
+            // TODO handle isSoftLogout parameter.
+            appCoroutineScope.launch(updateTokensDispatcher) {
+                val currentClient = this@RustClientSessionDelegate.client
+                if (currentClient == null) {
+                    clientLog.w("didReceiveAuthError -> no client, exiting")
+                    return@launch
+                }
+                val existingData = sessionStore.getSession(currentClient.sessionId.value)
+                val (anonymizedAccessToken, anonymizedRefreshToken) = existingData.anonymizedTokens()
+                clientLog.d(
+                    "Removing session data with access token '$anonymizedAccessToken' " +
+                        "and refresh token '$anonymizedRefreshToken'."
+                )
+                if (existingData != null) {
+                    // Set isTokenValid to false
+                    val newData = existingData.copy(isTokenValid = false)
+                    sessionStore.updateData(newData)
+                    clientLog.d("Invalidated session data with access token: '$anonymizedAccessToken'.")
+                } else {
+                    clientLog.d("No session data found.")
+                }
+                client?.logout(ignoreSdkError = true, forced = true)
+            }.invokeOnCompletion {
+                if (it != null) {
+                    clientLog.e(it, "Failed to remove session data.")
+                }
+            }
+        } else {
+            clientLog.v("didReceiveAuthError -> already cleaning up")
+        }
+    }
+
+    override fun didRefreshTokens() {
+        // This is done in `saveSessionInKeychain(Session)` instead.
+    }
+
+    override fun retrieveSessionFromKeychain(userId: String): Session {
+        // This should never be called, as it's only used for multi-process setups
+        error("retrieveSessionFromKeychain should never be called for Android")
```

Comment:
Maybe do not crash the app but just log an error?

Reply:
Then I need to create a fake `Session` object, or retrieve it from disk so it conforms to the method signature.